solr-http-retry.go

```diff
@@ -0,0 +1,80 @@
+package solr
+
+import (
+	"log"
+	"net/url"
+	"time"
+)
+
+type SolrHttpRetrier struct {
+	solrCli            SolrHTTP
+	retries            int
+	exponentialBackoff time.Duration
+	timeout            time.Duration
+}
+
+func NewSolrHttpRetrier(solrHttp SolrHTTP, retries int, exponentialBackoff time.Duration, timeout time.Duration) SolrHTTP {
+
+	solrRetrier := SolrHttpRetrier{solrCli: solrHttp, retries: retries, exponentialBackoff: exponentialBackoff, timeout: timeout}
+	return &solrRetrier
+}
+
+func (s *SolrHttpRetrier) Read(opts ...func(url.Values)) (SolrResponse, error) {
+	now := time.Now()
+	var resp SolrResponse
+	var err error
+	for attempt := 0; attempt < s.retries; attempt++ {
+		resp, err = s.solrCli.Read(opts...)
+		if err != nil {
+			s.Logger().Printf("[Solr Http Retrier] Error Retrying %v ", err)
+			if !s.backoff(now, attempt) {
+				s.Logger().Println("[Solr Http Retrier] Timeout exceeded", err)
+				break
+			}
+			continue
+		}
+		break
+	}
+	return resp, err
+}
+
+func (s *SolrHttpRetrier) Update(docID string, updateOnly bool, doc interface{}, opts ...func(url.Values)) error {
+	now := time.Now()
+	var err error
+	for attempt := 0; attempt < s.retries; attempt++ {
+		err := s.solrCli.Update(docID, updateOnly, doc, opts...)
+		if err != nil {
+			log.Printf("[Solr Http Retrier] Error Retrying %v ", err)
+			if !s.backoff(now, attempt) {
+				log.Println("[Solr Http Retrier] Timeout exceeded", err)
+				break
+			}
+			continue
+		}
+		break
+	}
+	return err
+}
+
+func (s *SolrHttpRetrier) Logger() Logger {
+	return s.solrCli.Logger()
+}
+
+//returns whether cap has been passed
+func (s *SolrHttpRetrier) backoff(now time.Time, attempt int) bool {
+	//cap the time, whichever is less
+	cap := min(s.timeout.Seconds(), s.exponentialBackoff.Seconds()*float64(attempt))
+	if time.Since(now)*time.Second < time.Duration(cap)*time.Second {
+		randomizedCap := time.Duration(cap) * time.Second
```

Comment:
removed the randomize, but left the var name by accident. randomizing will restagger the connections to prevent thundering heard issues